File: examples/diffie_hellman_test.ts
import { assertEquals } from "@std/assert";
import { simulate } from "../typescript/simulator.ts";
import {
  DiffieHellman_A,
  DiffieHellman_B,
  Math,
  Secret_A,
  Secret_B,
} from "./diffie_hellman/diffie_hellman.ts";
import { Env } from "../typescript/runtime.ts";

Deno.test("simulate diffie hellman", async () => {
  const mathImpl: Math = {
    Exp(_: Env, base: number, exp: number): Promise<number> {
      return Promise.resolve(base ** exp);
    },
  };

  const result = await simulate({
    A: async (env) => {
      return await DiffieHellman_A(env, mathImpl);
    },
    B: async (env) => {
      return await DiffieHellman_B(env, mathImpl);
    },
  });

  assertEquals(result, {
    A: {
      receives: [{ sender: "B", value: 10 }],
      return: new Secret_A({ A: 18 }),
      sends: [{ receivers: ["B"], value: 4 }],
    },
    B: {
      receives: [{ sender: "A", value: 4 }],
      return: new Secret_B({ B: 18 }),
      sends: [{ receivers: ["A"], value: 10 }],
    },
  });
});

File: examples/diffie_hellman_test.go
package examples_test

import (
	"math"
	"testing"

	"github.com/tempo-lang/tempo/examples/diffie_hellman"
	"github.com/tempo-lang/tempo/runtime"
	"github.com/tempo-lang/tempo/simulator"
	"github.com/tempo-lang/tempo/transports"

	"github.com/google/go-cmp/cmp"
)

type mathImpl struct{}

// Exp implements diffie_hellman.Math.
func (m *mathImpl) Exp(env *runtime.Env, base int, exp int) int {
	return int(math.Pow(float64(base), float64(exp)))
}

func TestDiffieHellmanSim(t *testing.T) {

	result := simulator.Run(
		simulator.Proc("A", func(env *runtime.Env) any {
			return diffie_hellman.DiffieHellman_A(env, &mathImpl{})
		}),
		simulator.Proc("B", func(env *runtime.Env) any {
			return diffie_hellman.DiffieHellman_B(env, &mathImpl{})
		}),
	)

	expected := []simulator.Result{
		{
			Return:   diffie_hellman.Secret_A{A: 18},
			Sends:    []transports.SendValue{{Value: int(4), Receivers: []string{"B"}}},
			Receives: []transports.RecvValue{{Value: int(10), Sender: "B"}},
		},
		{
			Return:   diffie_hellman.Secret_B{B: 18},
			Sends:    []transports.SendValue{{Value: int(10), Receivers: []string{"A"}}},
			Receives: []transports.RecvValue{{Value: int(4), Sender: "A"}},
		},
	}

	if diff := cmp.Diff(result, expected); diff != "" {
		t.Errorf("DiffieHellman result did not match expected:\n%s", diff)
	}
}
